### DISCUSSION TITLE
How do I hook into the startup of a NextJS server so that I can do one-off initialization?

### DISCUSSION THREAD

**Discussion by sarowarhosen03** [vercel/next.js#50198] [Help] [answered]
### Summary

i want to run some function or execute  something which is run when the app is started where i should write the startup code ? i express js we write in the app.js  or insdex.js? i need run the codes just one time when app is started

**Comment by icyJoseph** [comment, marked as answer]
## A long awaited feature

There's a new way implemented by the Next.js team. If you have an `instrumentation` file, its contents will be run on server start!

> It has the experimental flag, so you gotta activate it in next.config.js
> It is also only made available recently, your Next.js version might not support it

- https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation

I think the biggest gotcha there, is that you gotta differentiate node.js from edge, because not all Node API's are available in the edge runtime:

```ts
export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    await import('./instrumentation-node');
  }
 
  if (process.env.NEXT_RUNTIME === 'edge') {
    await import('./instrumentation-edge');
  }
}
```

It also supports TypeScript!


## Otherwise, up to next@13.1.2

Well, you can use the `next.config.js` file in function form:

```js
const nextConfig = {
  reactStrictMode: true,
};

module.exports = () => {
  console.log("Loading Config");
  return nextConfig;
};
```

This one runs at exactly three points in time, dev server start, build process start, and production server start. You must copy it to wherever you run the production server!

**Reply by sarowarhosen03** [reply]
thank you so much finally I am got it.

**Reply by ODreelist** [reply]
> This one runs at exactly three points in time, dev server start, build process start, and production server start. You must copy it to wherever you run the production server!

is this still true and is this pattern still supported in next14? Seems to try to execute multiple times when running next dev

**Comment by sarowarhosen03** [comment]
![Screenshot from 2023-05-24 11-49-20](https://github.com/vercel/next.js/assets/98280301/076cce12-cdd1-4e3a-8eb0-1a4e4990e91d)
but here is an issu its run tow times

**Reply by icyJoseph** [reply]
I see, something must have changed in how the next.config.js is loaded, it used to load only once :/ 

It worked up to 13.1.2 - with anything after that loads it twice. 

You should use the `instrumentation` file instead.

**Reply by sarowarhosen03** [reply]
instrumentation file in the root of my project it's not work i think I have some mistake can you please give me a demo with Instrumentation file

**Reply by icyJoseph** [reply]
Did you activate the experimental flag in `next.config.js`? And have you got the latest next.js version?

**Reply by sarowarhosen03** [reply]
I don't know how to activate the expatimsntel flag in next.config.js

**Reply by icyJoseph** [reply]
Inside the next.config.js file you should have this:

```js

const nextConfig = {
/* Other config goes here*/
  experimental : {
     instrumentationHook: true
   }
}

module.exports = nextConfig

```

**Reply by sarowarhosen03** [reply]
thank you so much 🥰

**Reply by sarowarhosen03** [reply]
my file structure is ![Screenshot_20230527-221531.jpg](https://github.com/vercel/next.js/assets/98280301/111e41c2-2c1e-4824-b77a-eab847b6fdff)

in my instrumentation.ts 

`export async function register() {
  console.log("hi from init")
}`
its okay. i couldn't see any console on startup

**Reply by icyJoseph** [reply]
Does it log if you open a page in the browser?

**Reply by sarowarhosen03** [reply]
> Does it log if you open a page in the browser? 

im  put the instrumentation.ts file in src folder and when am visit first time the message is run tow time. I want that the instrumentation.ts run in one time when the server is started.  when am use `module.exports = () => {
  console.log("Loading Config");
  return nextConfig;
}; `
the callback run tow time when server is run . i can handle it by putting condition but if nextjs make sure the nextjs.config callback run in one time when server is started or instrumentation.ts run one time it would better

**Reply by icyJoseph** [reply]
Yes but the instrumentation file runs per environment:

```ts
export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    await import('./instrumentation-node');
  }
 
  if (process.env.NEXT_RUNTIME === 'edge') {
    await import('./instrumentation-edge');
  }
}
```

So you probably just want to check for `NEXT _RUNTIME` equal to `"nodejs"`

Refer to the docs too: https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation
